File: typescript/libs/workspace/adapters/json/system/src/lib/session/Strategy.ts
import { IObject, IStrategy } from '@allors/workspace/domain/system';
import { AssociationType, Class, MethodType, Origin, RoleType } from '@allors/workspace/meta/system';
import { DatabaseRecord } from '../database/DatabaseRecord';
import { DatabaseOriginState, InitialVersion, UnknownVersion } from './originstate/DatabaseOriginState';
import { WorkspaceOriginState } from './originstate/WorkspaceOriginState';
import { isNewId, Session } from './Session';
import { Numbers, enumerate } from '../collections/Numbers';

export class Strategy implements IStrategy {
  DatabaseOriginState: DatabaseOriginState;
  WorkspaceOriginState: WorkspaceOriginState;

  private _object: IObject;

  constructor(public session: Session, public cls: Class, public id: number) {
    if (this.cls.origin !== Origin.Session) {
      this.WorkspaceOriginState = new WorkspaceOriginState(this, this.session.workspace.getRecord(this.id));
    }
  }

  static fromDatabaseRecord(session: Session, databaseRecord: DatabaseRecord) {
    return new Strategy(session, databaseRecord.cls, databaseRecord.id);
  }

  public get version(): number {
    switch (this.cls.origin) {
      case Origin.Session:
        return InitialVersion;
      case Origin.Workspace:
        return this.WorkspaceOriginState.Version;
      case Origin.Database:
        return this.DatabaseOriginState.Version;
      default:
        throw new Error('Unknown origin');
    }
  }

  public get isNew(): boolean {
    return isNewId(this.id);
  }

  public get object(): IObject {
    return (this._object ??= this.session.workspace.database.configuration.objectFactory.create(this));
  }

  public exist(roleType: RoleType): boolean {
    if (roleType.objectType.isUnit) {
      return this.getUnit(roleType) != null;
    }

    if (roleType.isOne) {
      return this.getComposite(roleType) != null;
    }

    return this.getComposites(roleType).Any();
  }

  public get(roleType: RoleType): any {
    if (roleType.objectType.isUnit) {
      return this.getUnit(roleType);
    }

    if (roleType.isOne) {
      return this.getComposite(roleType);
    }

    return this.getComposites(roleType);
  }

  public getUnit(roleType: RoleType): any {
    switch (this.cls.origin) {
      case Origin.Session:
        return this.session.GetRole(this, roleType);
      case Origin.Workspace:
        return this.WorkspaceOriginState?.GetRole(roleType);
      case Origin.Database:
        if (this.DatabaseOriginState) {
          if (this.DatabaseOriginState.Version === UnknownVersion) {
            throw new Error('Uknown version');
          }

          if (this.canRead(roleType)) {
            return this.DatabaseOriginState?.GetRole(roleType);
          }
        }

        return null;
      default:
        throw new Error('Unknown origin');
    }
  }

  public getComposite(roleType: RoleType) {
    switch (this.cls.origin) {
      case Origin.Session:
        return this.session.GetRole(this, roleType);
      case Origin.Workspace:
        return this.WorkspaceOriginState?.GetRole(roleType);
      case Origin.Database:
        if (this.DatabaseOriginState) {
          if (this.DatabaseOriginState.Version === UnknownVersion) {
            throw new Error('Uknown version');
          }

          if (this.canRead(roleType)) {
            return this.DatabaseOriginState?.GetRole(roleType);
          }
        }

        return null;
      default:
        throw new Error('Unknown origin');
    }
  }

  public *getComposites<T>(roleType: RoleType): Generator<T, void, unknown> {
    let roles: IObject[];

    switch (this.cls.origin) {
      case Origin.Session:
        roles = this.session.GetRole(this, roleType) as IObject[];
        break;
      case Origin.Workspace:
        roles = this.WorkspaceOriginState?.GetRole(roleType) as IObject[];
        break;
      case Origin.Database:
        if (this.DatabaseOriginState) {
          if (this.DatabaseOriginState.Version === UnknownVersion) {
            throw new Error('Uknown version');
          }

          if (this.canRead(roleType)) {
            roles = this.DatabaseOriginState?.GetRole(roleType) as IObject[];
          }
        }

        break;
      default:
        throw new Error('Unknown origin');
    }

    if (roles != null) {
      for (const role of roles) {
        yield role;
      }
    }
  }

  public set(roleType: RoleType, value: any) {
    if (roleType.objectType.isUnit) {
      this.setUnit(roleType, value);
    } else if (roleType.isOne) {
      this.setComposite(roleType, value);
    } else {
      this.setComposites(roleType, value);
    }
  }

  public setUnit(roleType: RoleType, value: any) {
    switch (roleType.origin) {
      case Origin.Session:
        this.session.sessionOriginState.SetUnitRole(this.id, roleType, value);
        break;
      case Origin.Workspace:
        this.WorkspaceOriginState?.SetUnitRole(roleType, value);
        break;
      case Origin.Database:
        if (this.canWrite(roleType)) {
          this.DatabaseOriginState?.SetUnitRole(roleType, value);
        }

        break;
      default:
        throw new Error('Unsupported Origin');
    }
  }

  public setComposite<T extends IObject>(roleType: RoleType, value: T) {
    switch (roleType.origin) {
      case Origin.Session:
        this.session.sessionOriginState.SetCompositeRole(this.id, roleType, value?.id);
        break;
      case Origin.Workspace:
        this.WorkspaceOriginState?.SetCompositeRole(roleType, value?.id);
        break;
      case Origin.Database:
        if (this.canWrite(roleType)) {
          this.DatabaseOriginState?.SetCompositeRole(roleType, value?.id);
        }

        break;
      default:
        throw new Error('Unsupported Origin');
    }
  }

  public setComposites(roleType: RoleType, role: IObject[]) {
    const roleNumbers = Numbers(role?.map((v) => v.id));

    switch (roleType.origin) {
      case Origin.Session:
        this.session.sessionOriginState.SetCompositesRole(this.id, roleType, roleNumbers);
        break;

      case Origin.Workspace:
        this.WorkspaceOriginState?.SetCompositesRole(roleType, roleNumbers);

        break;

      case Origin.Database:
        if (this.canWrite(roleType)) {
          this.DatabaseOriginState?.SetCompositesRole(roleType, roleNumbers);
        }

        break;
      default:
        throw new Error('Unsupported Origin');
    }
  }

  public add(roleType: RoleType, value: any) {
    switch (roleType.origin) {
      case Origin.Session:
        this.session.sessionOriginState.AddRole(this.id, roleType, value.Id);
        break;
      case Origin.Workspace:
        this.WorkspaceOriginState.AddCompositeRole(roleType, value.Id);
        break;
      case Origin.Database:
        if (this.canWrite(roleType)) {
          this.DatabaseOriginState.AddCompositeRole(roleType, value.Id);
        }

        break;
      default:
        throw new Error('Unsupported Origin');
        break;
    }
  }

  public remove(roleType: RoleType, value: any) {
    switch (roleType.origin) {
      case Origin.Session:
        this.session.sessionOriginState.AddRole(this.id, roleType, value.Id);
        break;
      case Origin.Workspace:
        this.WorkspaceOriginState.RemoveCompositeRole(roleType, value.Id);
        break;
      case Origin.Database:
        if (this.canWrite(roleType)) {
          this.DatabaseOriginState.RemoveCompositeRole(roleType, value.Id);
        }

        break;
      default:
        throw new Error('Unsupported Origin');
    }
  }

  public removeAll(roleType: RoleType) {
    if (roleType.objectType.isUnit) {
      this.setUnit(roleType, null);
    } else if (roleType.isOne) {
      this.setComposite(roleType, null);
    } else {
      this.setComposites(roleType, null);
    }
  }

  public getCompositeAssociation(associationType: AssociationType) {
    if (associationType.origin != Origin.Session) {
      return this.session.GetCompositeAssociation(this.id, associationType);
    }

    let association = this.session.sessionOriginState.Get(this.id, associationType);
    return association != null ? this.session.getOne(association) : null;
  }

  public *getCompositesAssociation<T>(associationType: AssociationType):  Generator<T, void, unknown> {
    if (associationType.origin != Origin.Session) {
      return this.session.GetCompositesAssociation(this.id, associationType);
    }

    const association = this.session.sessionOriginState.Get(this.id, associationType);

    for (const id of enumerate(association)) {
      yield this.session.getOne(id);
    }
  }

  public canRead(roleType: RoleType): boolean {
    return this.DatabaseOriginState?.CanRead(roleType) ?? true;
  }

  public canWrite(roleType: RoleType): boolean {
    return this.DatabaseOriginState?.CanWrite(roleType) ?? true;
  }

  public canExecute(methodType: MethodType): boolean {
    return this.DatabaseOriginState?.CanExecute(methodType) ?? false;
  }

  public isAssociationForRole(roleType: RoleType, forRoleId: number): boolean {
    const role = this.session.sessionOriginState.Get(this.id, roleType);

    switch (roleType.origin) {
      case Origin.Session:
        return role === forRoleId;
      case Origin.Workspace:
        return this.WorkspaceOriginState?.IsAssociationForRole(roleType, forRoleId) ?? false;
      case Origin.Database:
        return this.DatabaseOriginState?.IsAssociationForRole(roleType, forRoleId) ?? false;
      default:
        throw new Error('Unsupported Origin');
    }
  }

  public onDatabasePushNewId(newId: number) {
    this.id = newId;
  }

  public onDatabasePushResponse(databaseRecord: DatabaseRecord) {
    this.DatabaseOriginState.PushResponse(databaseRecord);
  }
}

File: dotnet/core/workspace/csharp/adapters/allors.workspace.adapters/session/Strategy.cs
// <copyright file="Object.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Linq;
    using Meta;

    public abstract class Strategy : IStrategy
    {
        private IObject @object;

        protected Strategy(Session session, IClass @class, long id)
        {
            this.Session = session;
            this.Id = id;
            this.Class = @class;

            if (!this.Class.HasSessionOrigin)
            {
                this.WorkspaceOriginState = new WorkspaceOriginState(this, this.Session.Workspace.GetRecord(this.Id));
            }
        }

        protected Strategy(Session session, DatabaseRecord databaseRecord)
        {
            this.Session = session;
            this.Id = databaseRecord.Id;
            this.Class = databaseRecord.Class;

            this.WorkspaceOriginState = new WorkspaceOriginState(this, this.Session.Workspace.GetRecord(this.Id));
        }

        public long Version =>
            this.Class.Origin switch
            {
                Origin.Session => Allors.Version.Initial.Value,
                Origin.Workspace => this.WorkspaceOriginState.Version,
                Origin.Database => this.DatabaseOriginState.Version,
                _ => throw new Exception()
            };

        public Session Session { get; }

        public DatabaseOriginState DatabaseOriginState { get; protected set; }

        public WorkspaceOriginState WorkspaceOriginState { get; }

        ISession IStrategy.Session => this.Session;

        public IClass Class { get; }

        public long Id { get; private set; }

        public bool IsNew => Session.IsNewId(this.Id);

        public IObject Object => this.@object ??= this.Session.Workspace.DatabaseConnection.Configuration.ObjectFactory.Create(this);

        public bool Exist(IRoleType roleType)
        {
            if (roleType.ObjectType.IsUnit)
            {
                return this.GetUnit(roleType) != null;
            }

            if (roleType.IsOne)
            {
                return this.GetComposite<IObject>(roleType) != null;
            }

            return this.GetComposites<IObject>(roleType).Any();
        }

        public object Get(IRoleType roleType)
        {
            if (roleType.ObjectType.IsUnit)
            {
                return this.GetUnit(roleType);
            }

            if (roleType.IsOne)
            {
                return this.GetComposite<IObject>(roleType);
            }

            return this.GetComposites<IObject>(roleType);
        }

        public object GetUnit(IRoleType roleType) =>
            roleType.Origin switch
            {
                Origin.Session => this.Session.GetRole(this, roleType),
                Origin.Workspace => this.WorkspaceOriginState?.GetRole(roleType),
                Origin.Database => this.DatabaseOriginState?.IsVersionUnknown == true ? throw new Exception() : this.CanRead(roleType) ? this.DatabaseOriginState?.GetRole(roleType) : null,
                _ => throw new ArgumentException("Unsupported Origin")
            };

        public T GetComposite<T>(IRoleType roleType) where T : IObject =>
            this.Session.GetOne<T>(roleType.Origin switch
            {
                Origin.Session => (long?)this.Session.GetRole(this, roleType),
                Origin.Workspace => (long?)this.WorkspaceOriginState?.GetRole(roleType),
                Origin.Database => this.DatabaseOriginState?.IsVersionUnknown == true ? throw new Exception() : this.CanRead(roleType) ? (long?)this.DatabaseOriginState?.GetRole(roleType) : null,
                _ => throw new ArgumentException("Unsupported Origin")
            });

        public IEnumerable<T> GetComposites<T>(IRoleType roleType) where T : IObject
        {
            var roles = roleType.Origin switch
            {
                Origin.Session => this.Session.GetRole(this, roleType),
                Origin.Workspace => this.WorkspaceOriginState?.GetRole(roleType),
                Origin.Database => this.DatabaseOriginState?.IsVersionUnknown == true ? throw new Exception() : this.CanRead(roleType) ? this.DatabaseOriginState?.GetRole(roleType) : null,
                _ => throw new ArgumentException("Unsupported Origin")
            };

            if (roles == null)
            {
                yield break;
            }

            foreach (var role in (IEnumerable<IObject>)roles)
            {
                yield return (T)role;
            }
        }

        public void Set(IRoleType roleType, object value)
        {
            if (roleType.ObjectType.IsUnit)
            {
                this.SetUnit(roleType, value);
            }
            else
            {
                if (roleType.IsOne)
                {
                    this.SetComposite(roleType, (IObject)value);
                }
                else
                {
                    this.SetComposites(roleType, (IEnumerable<IObject>)value);
                }
            }
        }

        public void SetUnit(IRoleType roleType, object value)
        {
            switch (roleType.Origin)
            {
                case Origin.Session:
                    this.Session.SessionOriginState.SetUnitRole(this.Id, roleType, value);
                    break;

                case Origin.Workspace:
                    this.WorkspaceOriginState?.SetUnitRole(roleType, value);

                    break;

                case Origin.Database:
                    if (this.CanWrite(roleType))
                    {
                        this.DatabaseOriginState?.SetUnitRole(roleType, value);
                    }

                    break;
                default:
                    throw new ArgumentException("Unsupported Origin");
            }
        }

        public void SetComposite<T>(IRoleType roleType, T value) where T : IObject
        {
            switch (roleType.Origin)
            {
                case Origin.Session:
                    this.Session.SessionOriginState.SetCompositeRole(this.Id, roleType, value?.Id);
                    break;

                case Origin.Workspace:
                    this.WorkspaceOriginState?.SetCompositeRole(roleType, value?.Id);

                    break;

                case Origin.Database:
                    if (this.CanWrite(roleType))
                    {
                        this.DatabaseOriginState?.SetCompositeRole(roleType, value?.Id);
                    }

                    break;
                default:
                    throw new ArgumentException("Unsupported Origin");
            }
        }

        public void SetComposites<T>(IRoleType roleType, in IEnumerable<T> role) where T : IObject
        {
            var roleNumbers = this.Session.Workspace.Numbers.From(role?.Select(v => v.Id));

            switch (roleType.Origin)
            {
                case Origin.Session:
                    this.Session.SessionOriginState.SetCompositesRole(this.Id, roleType, roleNumbers);
                    break;

                case Origin.Workspace:
                    this.WorkspaceOriginState?.SetCompositesRole(roleType, roleNumbers);

                    break;

                case Origin.Database:
                    if (this.CanWrite(roleType))
                    {
                        this.DatabaseOriginState?.SetCompositesRole(roleType, roleNumbers);
                    }

                    break;
                default:
                    throw new ArgumentException("Unsupported Origin");
            }
        }

        public void Add<T>(IRoleType roleType, T value) where T : IObject
        {
            switch (roleType.Origin)
            {
                case Origin.Session:
                    this.Session.SessionOriginState.AddRole(this.Id, roleType, value.Id);
                    break;

                case Origin.Workspace:
                    this.WorkspaceOriginState.AddCompositeRole(roleType, value.Id);

                    break;

                case Origin.Database:
                    if (this.CanWrite(roleType))
                    {
                        this.DatabaseOriginState.AddCompositeRole(roleType, value.Id);
                    }

                    break;
                default:
                    throw new ArgumentException("Unsupported Origin");
            }
        }

        public void Remove<T>(IRoleType roleType, T value) where T : IObject
        {
            switch (roleType.Origin)
            {
                case Origin.Session:
                    this.Session.SessionOriginState.AddRole(this.Id, roleType, value.Id);
                    break;

                case Origin.Workspace:
                    this.WorkspaceOriginState.RemoveCompositeRole(roleType, value.Id);

                    break;

                case Origin.Database:
                    if (this.CanWrite(roleType))
                    {
                        this.DatabaseOriginState.RemoveCompositeRole(roleType, value.Id);
                    }

                    break;
                default:
                    throw new ArgumentException("Unsupported Origin");
            }
        }

        public void RemoveAll(IRoleType roleType)
        {
            if (roleType.ObjectType.IsUnit)
            {
                this.SetUnit(roleType, null);
            }
            else
            {
                if (roleType.IsOne)
                {
                    this.SetComposite(roleType, (IObject)null);
                }
                else
                {
                    this.SetComposites(roleType, (IEnumerable<IObject>)null);
                }
            }
        }

        public T GetCompositeAssociation<T>(IAssociationType associationType) where T : IObject
        {
            if (associationType.Origin != Origin.Session)
            {
                return this.Session.GetCompositeAssociation<T>(this.Id, associationType);
            }

            var association = (long?)this.Session.SessionOriginState.Get(this.Id, associationType);
            return association != null ? this.Session.GetOne<T>(association) : default;
        }

        public IEnumerable<T> GetCompositesAssociation<T>(IAssociationType associationType) where T : IObject
        {
            if (associationType.Origin != Origin.Session)
            {
                return this.Session.GetCompositesAssociation<T>(this.Id, associationType);
            }

            var association = this.Session.SessionOriginState.Get(this.Id, associationType);

            return association switch
            {
                long id => new[] { this.Session.GetOne<T>(id) },
                long[] ids => ids.Select(v => this.Session.GetOne<T>(v)).ToArray(),
                _ => Array.Empty<T>()
            };
        }

        public bool CanRead(IRoleType roleType) => this.DatabaseOriginState?.CanRead(roleType) ?? true;

        public bool CanWrite(IRoleType roleType) => this.DatabaseOriginState?.CanWrite(roleType) ?? true;

        public bool CanExecute(IMethodType methodType) => this.DatabaseOriginState?.CanExecute(methodType) ?? false;

        public bool IsAssociationForRole(IRoleType roleType, long forRoleId)
        {
            var role = this.Session.SessionOriginState.Get(this.Id, roleType);
            return roleType.Origin switch
            {
                Origin.Session => Equals(role, forRoleId),
                Origin.Workspace => this.WorkspaceOriginState?.IsAssociationForRole(roleType, forRoleId) ?? false,
                Origin.Database => this.DatabaseOriginState?.IsAssociationForRole(roleType, forRoleId) ?? false,
                _ => throw new ArgumentException("Unsupported Origin")
            };
        }

        public void OnDatabasePushNewId(long newId) => this.Id = newId;

        public void OnDatabasePushResponse(DatabaseRecord databaseRecord) => this.DatabaseOriginState.PushResponse(databaseRecord);
    }
}
